### PR TITLE
WIP: screen view search

### PR DESCRIPTION
**Pull request by PiniAmiram**

### CODE REVIEW

File: src/app/@ansyn/ansyn/modules/core/components/context-menu/context-menu.component.ts [hunk 1]
```diff
@@ -96,6 +97,10 @@ export class ContextMenuComponent implements OnInit {
 		tap<any>((region) => this.geoFilter = region.type)
 	);
 
+	screenViewMode$ = this.store.select(selectGeoFilterStatus).pipe(
+		map((geoFilter: IGeoFilterStatus) => geoFilter.type === 'ScreenView')
```
Comment (sulimani55) [inline src/app/@ansyn/ansyn/modules/core/components/context-menu/context-menu.component.ts:101]: use CaseGeoFilters.ScreenViews

File: src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts [hunk 2]
```diff
@@ -321,10 +334,40 @@ export class MapAppEffects {
...
 
+	@Effect({ dispatch: false })
+	searchByExtentPolygon$: Observable<any> = this.actions$.pipe(
+		ofType(MapActionTypes.POSITION_CHANGED, MapActionTypes.SET_ACTIVE_MAP_ID),
+		withLatestFrom(this.store$.select(selectMaps), this.store$.select(selectActiveMapId), this.store$.select(selectGeoFilterStatus)),
+		map(([action, mapList, activeMapId, geoFilterStatus]) => {
+			const activeMap: IMapSettings = mapList[activeMapId];
+			return [activeMap.data.position, geoFilterStatus];
+		}),
+		filter(([position, geoFilterStatus]: [IImageryMapPosition, IGeoFilterStatus]) => Boolean(position) && geoFilterStatus.type === 'ScreenView'),
```
Comment (sulimani55) [inline src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts:345]: use CaseGeoFilter.ScreenView

File: src/app/@ansyn/ansyn/modules/plugins/openlayers/plugins/visualizers/region/region.visualizer.ts [hunk 3]
```diff
@@ -58,16 +58,16 @@ export abstract class RegionVisualizer extends EntitiesVisualizer {
...
 	);
 
 	constructor(public store$: Store<any>, public actions$: Actions, public projectionService: OpenLayersProjectionService, public geoFilter: CaseGeoFilter) {
 		super();
 	}
 
-	drawChanges([region, regionAsVisualizer, isActive, isMinimalistViewMode]) {
-		if (regionAsVisualizer && !isActive && !isMinimalistViewMode) {
+	drawChanges([geoFilterType, region, regionAsVisualizer, isActive, isMinimalistViewMode]) {
+		if (regionAsVisualizer && !isActive && !isMinimalistViewMode && geoFilterType !== 'ScreenView') {
```
Comment (sulimani55) [inline src/app/@ansyn/ansyn/modules/plugins/openlayers/plugins/visualizers/region/region.visualizer.ts:70]: use CaseGeoFilter.ScreenView

**Review by sulimani55 — APPROVED** [review]

File: src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts [hunk 4]
```diff
@@ -321,10 +334,40 @@ export class MapAppEffects {
...
+			const activeMap: IMapSettings = mapList[activeMapId];
+			return [activeMap.data.position, geoFilterStatus];
+		}),
+		filter(([position, geoFilterStatus]: [IImageryMapPosition, IGeoFilterStatus]) => Boolean(position) && geoFilterStatus.type === CaseGeoFilter.ScreenView),
+		debounceTime(VIEW_SEARCH_DEBOUNCE_TIME),
+		tap(([position, geoFilterStatus]: [IImageryMapPosition, IGeoFilterStatus]) => {
+			let zoom = position.projectedState.zoom;
+
+			if (!equalPolygons(position.extentPolygon, extentPolygon)) {
+				if (zoom < VIEW_SEARCH_ZOOM_LIMIT) {
```
Comment (asafMasa) [inline src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts:351]: won't work on overlays
you can't use zoom on images only resolution - try to use the scale - or calculate BBOX width size in meters

File: src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts [hunk 5]
```diff
@@ -85,8 +88,18 @@ import {
...
 } from '../../modules/overlays/overlay-status/config/overlay-status-config';
 import { MeasureDistanceVisualizer } from '../../modules/plugins/openlayers/plugins/visualizers/tools/measure-distance.visualizer';
+import { IGeoFilterStatus, selectGeoFilterStatus } from '../../modules/status-bar/reducers/status-bar.reducer';
+import { Polygon } from '@turf/turf';
+import { UpdateGeoFilterStatus } from '../../modules/status-bar/actions/status-bar.actions';
+import { ScreenViewSearchVisualizer } from '../../modules/plugins/openlayers/plugins/visualizers/region/screen-view.visualiser';
 
 const FOOTPRINT_INSIDE_MAP_RATIO = 1;
+const VIEW_SEARCH_ZOOM_LIMIT = 14;
+const VIEW_SEARCH_DEBOUNCE_TIME = 1000;
```
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts:98]: config for meters

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Review by Danaorlov1005 — APPROVED** [review]

File: src/app/@ansyn/ansyn/app-effects/effects/status-bar.app.effects.ts [hunk 6]
```diff
@@ -93,7 +90,7 @@ export class StatusBarAppEffects {
 	onCancelGeoFilter$ = this.actions$.pipe(
 		ofType<UpdateGeoFilterStatus>(StatusBarActionsTypes.UPDATE_GEO_FILTER_STATUS),
 		filter(action => action.payload === undefined),
-		withLatestFrom(this.store.select(selectRegion)),
+		withLatestFrom(this.store.select(selectGeoFilterStatus)),
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/app-effects/effects/status-bar.app.effects.ts:93]: it should be taken from the region.
to see why,
1. draw a polygon search.
2. check the point search
3. click on the menu.

the polygon search redraw on the map, but in the geoFilter will be write `Point`

File: src/app/@ansyn/ansyn/modules/core/components/context-menu/context-menu.component.less [hunk 7]
```diff
@@ -61,6 +61,17 @@
 					font-size: 40px;
 					font-weight: bold;
 				}
+
+				&:disabled {
+					&:hover {
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/modules/core/components/context-menu/context-menu.component.less:66]: on disable make the button always gray not only on hover.

File: src/app/@ansyn/ansyn/modules/plugins/openlayers/plugins/visualizers/region/screen-view.visualiser.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { Actions } from '@ngrx/effects';
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/modules/plugins/openlayers/plugins/visualizers/region/screen-view.visualiser.ts:1]: why screen-view need visualiser?

File: src/app/@ansyn/imagery/utils/geo.ts [hunk 9]
```diff
@@ -128,6 +128,11 @@ export function polygonsDontIntersect(extentPolygon, footprint, overlayCoverage)
 	return intersection < overlayCoverage;
 }
 
+export function equalPolygons(firstPolygon: Polygon, secondPolygon: Polygon): boolean {
+	const equal = difference(firstPolygon, secondPolygon) === null && difference(secondPolygon, firstPolygon) === null;
```
Comment (tltk90) [inline src/app/@ansyn/imagery/utils/geo.ts:132]: you can use booleanEqual from turf.

**Review by tltk90 — CHANGES_REQUESTED** [review]

File: src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts [hunk 10]
```diff
@@ -61,7 +62,20 @@ export class SelectCaseAppEffects {
 		// map
 		const { data } = state.maps;
 		// context
-		const { favoriteOverlays, region, dataInputFilters, miscOverlays } = state;
+		const { favoriteOverlays, dataInputFilters, miscOverlays } = state;
+
+		let region: Feature<Polygon | Point>;
+		if (state.region.type !== "Feature") {
+			region = {
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.ts:69]: use turf `feature` function to create a feature.

File: src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.spec.ts [hunk 11]
```diff
@@ -133,7 +139,7 @@ describe('SelectCaseAppEffects', () => {
 			b: new SetActiveMapId(state.maps.activeMapId),
 			c: new SetLayoutAction(<any>maps.layout),
 			d: new SetOverlaysCriteriaAction({ time, region, dataInputFilters }, { noInitialSearch }),
-			e: new UpdateGeoFilterStatus({active: false, type: region.type}),
+			e: new UpdateGeoFilterStatus({active: false, type: region.geometry.type}),
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/app-effects/effects/cases/select-case.app.effects.spec.ts:142]: region.properites.searchMode

File: src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts [hunk 12]
```diff
@@ -457,6 +493,17 @@ export class MapAppEffects {
 		return EMPTY;
 	}
 
+	createRegion(polygon) {
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/app-effects/effects/map.app.effects.ts:496]: use turf `feature` instead.

File: src/app/@ansyn/ansyn/modules/plugins/openlayers/plugins/visualizers/region/pin-point.visualizer.ts [hunk 13]
```diff
@@ -38,19 +39,31 @@ export class PinPointVisualizer extends RegionVisualizer {
...
-		return geoJsonFeature.geometry;
+		geoJsonFeature.properties = {
+			searchMode: "Point"
+		}
+		return geoJsonFeature;
 	}
 
 	onContextMenu(coordinates: Position): void {
-		const region = turf.geometry('Point', coordinates);
+		const region: Feature<Point> = {
```
Comment (tltk90) [inline src/app/@ansyn/ansyn/modules/plugins/openlayers/plugins/visualizers/region/pin-point.visualizer.ts:57]: use turf `feature` instead

**Review by tltk90 — CHANGES_REQUESTED** [review]
💯

**Review by tltk90 — APPROVED** [review]
👍

**Pull request merged**